### PR TITLE
ACR support

### PR DESCRIPTION
**Pull request by Eskibear**

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -122,6 +129,25 @@ export function activate(context: vscode.ExtensionContext) {
         Executor.onDidCloseTerminal(closedTerminal);
     }));
 
+    context.subscriptions.push(vscode.commands.registerCommand("docker-explorer.pullImage", (tagItem) => {
```
Comment (formulahendry) [inline src/extension.ts]: Track AI event for all commands.

File: src/azureContainerRegistries.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            Executor.runInTerminal(`az group create --name ${resourceGroup} --location ${location}`);
+            // Create an App Service Plan
+            Executor.runInTerminal(`az appservice plan create --name ${appServicePlan} --resource-group ${resourceGroup} --location ${location} --is-linux --sku S1`);
+            // Create a Web App
+            Executor.runInTerminal(`az webapp create --name ${appName} --plan ${appServicePlan} --resource-group ${resourceGroup}`);
+            // Configure Web App with a Custom Docker Container from Docker Hub
+            Executor.runInTerminal(`az webapp config container set --docker-custom-image-name ${dockerContainerPath} --docker-registry-server-password ${password} --docker-registry-server-url ${serverUrl} --docker-registry-server-user ${registry} --name ${appName} --resource-group ${resourceGroup}`);
+
+            // Sync output Web URL
+            Executor.runInTerminal(`echo "Web App is deployed to http://${appName}.azurewebsites.net/"`);
```
Comment (formulahendry) [inline src/azureContainerRegistries.ts]: 机智

File: src/azureContainerRegistries.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            const credentialObj = JSON.parse(credential);
+            const password = credentialObj.passwords[0].value;
+            const loginResult = Executor.execSync(`docker login ${registry}.azurecr.io -u ${registry} -p ${password}`);
+            if (loginResult.indexOf("Login Succeeded") >= 0) {
+                Executor.runInTerminal(`docker pull ${registry}.azurecr.io/${repository}:${tag}`);
+            }
+        }
+    }
+
+    public removeImage(tag: string, repository: string, registry: string): void {
```
Comment (formulahendry) [inline src/azureContainerRegistries.ts]: Is it used?

File: package.json [hunk 4]
```diff
@@ -121,6 +138,26 @@
                 "title": "Remove"
             },
             {
+                "command": "docker-explorer.pushImage",
+                "title": "Push"
```
Comment (formulahendry) [inline package.json]: Push to ACR

File: src/azureContainerRegistries.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    public logout() {
+        Executor.runInTerminal(`az logout`);
+    }
+
+    public getTreeItem(element: ACRTreeItem): vscode.TreeItem {
+        return element;
+    }
+
+    public getChildren(element?: ACRTreeItem): Thenable<ACRTreeItem[]> {
+        if (element === undefined) {
```
Comment (formulahendry) [inline src/azureContainerRegistries.ts]: Error handling. What if user has not run 'az login'?

File: src/extension.ts [hunk 6]
```diff
@@ -122,6 +129,25 @@ export function activate(context: vscode.ExtensionContext) {
...
+        azureContainerRegistries.deployToAzure(tagItem.node.name, tagItem.node.parent.name, tagItem.node.parent.parent.name);
+    }));
+    context.subscriptions.push(vscode.commands.registerCommand("docker-explorer.loginAzureCli", () => {
+        azureContainerRegistries.login();
+    }));
+
+    context.subscriptions.push(vscode.commands.registerCommand("docker-explorer.logoutAzureCli", () => {
+        azureContainerRegistries.logout();
+    }));
+    context.subscriptions.push(vscode.commands.registerCommand("docker-explorer.refreshAzureRegistries", () => {
```
Comment (formulahendry) [inline src/extension.ts]: leave blank line

File: src/extension.ts (hunk 1 above)
Comment (Eskibear) [inline src/extension.ts]: done.

File: src/azureContainerRegistries.ts (hunk 5 above)
Comment (Eskibear) [inline src/azureContainerRegistries.ts]: added.

File: src/extension.ts [hunk 7]
```diff
@@ -131,18 +131,22 @@ export function activate(context: vscode.ExtensionContext) {
...
         azureContainerRegistries.pullImage(tagItem.node.name, tagItem.node.parent.name, tagItem.node.parent.parent.name);
+        AppInsightsClient.sendEvent("pullImage");
     }));
 
     context.subscriptions.push(vscode.commands.registerCommand("docker-explorer.deployToAzure", (tagItem) => {
         azureContainerRegistries.deployToAzure(tagItem.node.name, tagItem.node.parent.name, tagItem.node.parent.parent.name);
+        AppInsightsClient.sendEvent("deployToAzure");
     }));
+
     context.subscriptions.push(vscode.commands.registerCommand("docker-explorer.loginAzureCli", () => {
```
Comment (formulahendry) [inline src/extension.ts]: Add AI event for login, logout as well

File: src/extension.ts [hunk 8]
```diff
@@ -131,18 +131,22 @@ export function activate(context: vscode.ExtensionContext) {
 
     context.subscriptions.push(vscode.commands.registerCommand("docker-explorer.pullImage", (tagItem) => {
         azureContainerRegistries.pullImage(tagItem.node.name, tagItem.node.parent.name, tagItem.node.parent.parent.name);
+        AppInsightsClient.sendEvent("pullImage");
```
Comment (formulahendry) [inline src/extension.ts]: pullImageFromACR

**Pull request merged**